File: Nov2020/17Nov/BalancedBrackets.cpp
/*
Problem Name: balanced brackets
Problem Link: https://www.hackerearth.com/practice/data-structures/stacks/basics-of-stacks/practice-problems/algorithm/balanced-brackets-3-4fc590c6/
*/

#include<bits/stdc++.h>
using namespace std;
int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0);             
    int n;
    cin>>n;
    while(n--){
        stack <char> b;
        string s;
        bool valid=true;
        cin>>s;
        for(int i=0;i<s.length();++i){
            if(s[i]=='(' || s[i]=='{' || s[i]=='['){
                b.push(s[i]);
            }
            else if(s[i]==')'  && b.size()>0 && b.top()=='('){
                b.pop();
            }
            else if(s[i]=='}'  && b.size()>0 && b.top()=='{'){
                b.pop();
            }
            else if(s[i]==']' && b.size()>0 && b.top()=='['){
                b.pop();
            }
            else{
                valid=false;
                break;
            }
        }
        if(!valid || b.empty()==false)  cout<<"NO"<<endl;
        else    cout<<"YES"<<endl;
    }
    return 0;
}

File: Nov2020/17Nov/BalancedBrackets.java
/*
Problem Name: balanced brackets
Problem Link: https://www.hackerearth.com/practice/data-structures/stacks/basics-of-stacks/practice-problems/algorithm/balanced-brackets-3-4fc590c6/
*/
import java.util.*;
class BalancedBrackets {
    public static void main(String args[] ) throws Exception {
        Scanner in = new Scanner(System.in);
        int t = in.nextInt();
        for(;t>0;t--){
            String s = in.next();
            Stack<Character> st = new Stack<Character>();
            for(char c: s.toCharArray()){
                if(st.isEmpty())
                    st.push(c);
                else if((st.peek() == '{') && (c == '}'))
                    st.pop();
                else if((st.peek() == '(') && (c == ')'))
                    st.pop();
                else if((st.peek() == '[') && (c == ']'))
                    st.pop();
                else
                    st.push(c);
            }
            if(st.isEmpty())
                System.out.println("YES");
            else
                System.out.println("NO");
        }
    }
}
